masknmf/utils/_serialization.py

```diff
@@ -0,0 +1,252 @@
+            _dicts_to_group(
+                h5file,
+                "{}{}/".format(path, key),
+                item.__dict__,
+                raise_meta_fail
+            )
+
+        else:
+            msg = "{} for item: {} not supported by HDF5" \
+                  "".format(type(item), item)
+
+            if raise_meta_fail:
+                raise TypeError(msg)
+
+            else:
+                h5file[path + key] = str(item)
+                warn("{}, storing whatever str(obj) returns"
+                     "".format(msg))
+
+
+def load_dict(filename, group):
+    """
+    Recursively load a dict from an hdf5 group in a file.
+
+    Parameters
+    ----------
+    filename : str
+        full path to the hdf5 file
+
+    group : str
+        Name of the group that contains the dict to load
+
+    Returns
+    -------
+    d : dict
+        dict loaded from the specified hdf5 group.
+    """
+
+    with h5py.File(filename, 'r') as h5file:
+        return _dicts_from_group(h5file, "{}/".format(group))
+
+
+def _dicts_from_group(h5file, path):
+    ans = {}
+    for key, item in h5file[path].items():
+        if "layout" in item.attrs:
+            # it's a torch tensor
+            match item.attrs["layout"]:
+                case "sparse_coo":
+                    # reconstruct sparse_coo
+                    indices = item["indices"][()]
+                    values = item["values"][()]
+                    size = item["size"][()]
+                    # needs to a tuple[int] for `torch.sparse_coo_tensor` to be happy
+                    size = tuple(size)
+
+                    ans[key] = torch.sparse_coo_tensor(
+                        indices=indices,
+                        values=values,
+                        size=size
+                    )
+
+                case "strided":
+                    ans[key] = torch.from_numpy(item[()])
+
+                case _:
+                    raise TypeError
+
+        elif isinstance(item, h5py.Dataset):
+            if item.attrs.__contains__('dtype'):
+                ans[key] = item[()].astype(item.attrs['dtype'])
+            else:
+                ans[key] = item[()]
+        elif isinstance(item, h5py.Group):
+            ans[key] = _dicts_from_group(h5file, path + key + '/')
+    return ans
+
+
+class Serializer:
+    # should specify set of serialized attributes
+    _serialized = {}
+
+    def _to_dict(self) -> dict:
+        d = {}
+
+        for key in self._serialized:
+            val = getattr(self, key)
+
+            # skip None so we don't have to deal with serializing them.
+            # HDF5 has no None type, and it can create headaches to properly
+            # serialize and de-serialize them. To get around this, any attributes
+            # that are None and required for de-serialization should be default None
+            # kwargs in the constructor of the class they are serialized from.
+            if val is None:
+                continue
```

Comment:
Important to note that any attribute specified in `_serialized` that are `None` will be skipped. If those attributes are required to be `None`, they must be a kwarg with a default value of `None`. For example: https://github.com/kushalkolar/masknmf-toolbox/blob/4672aa23ccb543f190f76212101d6d7c9106e401/masknmf/compression/pmd_array.py#L119